src/plugins/internalnotification/internalnotification.c

```diff
@@ -0,0 +1,263 @@
+				KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/open", KEY_FUNC, elektraInternalnotificationOpen,
+				KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/close", KEY_FUNC, elektraInternalnotificationClose,
+				KEY_END),
+
+			// Export registerInt function and required plugin handle
+			keyNew ("system/elektra/modules/internalnotification/exports/elektraInternalnotificationRegisterInt", KEY_FUNC,
+				elektraInternalnotificationRegisterInt, KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/handle", KEY_BINARY, KEY_SIZE, sizeof handle,
+				KEY_VALUE, &handle, KEY_END),
+
+#include ELEKTRA_README (internalnotification)
+
+			keyNew ("system/elektra/modules/internalnotification/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END), KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		return 1;
+	}
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+	fprintf (stderr, "elektraInternalnotificationGet keyName=%s\n", parentKeyName);
+#endif
+
+	ListPointer * listPointer = elektraPluginGetData (handle);
+	elektraInternalnotificationUpdateRegisteredKeys (listPointer, returned);
+
+	return 1;
+}
+
+int elektraInternalnotificationSet (Plugin * handle, KeySet * returned, Key * parentKey)
+{
+
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+	fprintf (stderr, "elektraInternalnotificationSet keyName=%s\n", keyName (parentKey));
+#endif
+
+	ListPointer * listPointer = elektraPluginGetData (handle);
+	elektraInternalnotificationUpdateRegisteredKeys (listPointer, returned);
+
+	return 1;
+}
+
+int elektraInternalnotificationOpen (Plugin * handle, Key * parentKey ELEKTRA_UNUSED)
+{
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+	fprintf (stderr, "elektraInternalnotificationOpen\n");
+#endif
+
+	ListPointer * listPointer = elektraPluginGetData (handle);
+	if (listPointer == NULL)
+	{
+		listPointer = elektraMalloc (sizeof *listPointer);
+		if (listPointer == NULL)
+		{
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+			fprintf (stderr, "elektraInternalnotificationOpen elektraMalloc failed!\n");
+#endif
+			return -1;
+		}
+		elektraPluginSetData (handle, listPointer);
+
+		// Inizialize list pointers for registered keys
+		listPointer->head = NULL;
+		listPointer->tail = NULL;
+	}
+
+	return 1;
+}
+
+int elektraInternalnotificationClose (Plugin * handle, Key * parentKey ELEKTRA_UNUSED)
+{
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+	fprintf (stderr, "elektraInternalnotificationClose\n");
+#endif
+
+	ListPointer * listPointer = elektraPluginGetData (handle);
+	if (listPointer != NULL)
+	{
+		// Free registrations
+		KeyRegistration * current = listPointer->head;
+		KeyRegistration * next;
+		while (current != NULL)
+		{
+			next = current->next;
+			elektraFree (current->name);
+			elektraFree (current);
+
+			current = next;
+		}
+
+		// Free list pointer
+		elektraFree (listPointer);
+		elektraPluginSetData (handle, NULL);
```

Comment:
free only frees the allocated memory, but the pointer still holds the address. i haven't read the whole code, but it's cleaner and allows you to make checks like elektraPluginGetData == NULL